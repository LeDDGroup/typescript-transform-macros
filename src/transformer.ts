import * as ts from "typescript";

interface Macro {
  name: ts.Identifier;
  value: ts.Expression;
}

function transformer<T extends ts.Node>(_program: ts.Program) {
  let counter = 0;
  return (context: ts.TransformationContext) => {
    const macros: Macro[] = [];
    return (node: T) => {
      const postExtractNode = ts.visitNode(node, extractMacros);
      return ts.visitNode(postExtractNode, resolveMacros);
    };
    function extractMacros(node: ts.Node): ts.Node | undefined {
      if (ts.isVariableStatement(node)) {
        const firstDeclaration = node.declarationList.declarations[0]; // TODO maybe check for more
        if (
          firstDeclaration.initializer &&
          ts.isCallExpression(firstDeclaration.initializer) &&
          ts.isIdentifier(firstDeclaration.initializer.expression) &&
          firstDeclaration.initializer.expression.text === "MACRO"
        ) {
          const name = firstDeclaration.name;
          if (!ts.isIdentifier(name)) {
            throw new Error(
              "Expected name to be Identifier for macro declaration"
            );
          }
          const value = firstDeclaration.initializer.arguments[0];
          macros.push({ name, value });
          return undefined;
        }
      }
      return ts.visitEachChild(node, extractMacros, context);
    }
    function resolveMacros(node: ts.Node): ts.Node | undefined {
      if (ts.isBlock(node)) {
        return ts.visitEachChild(
          ts.updateBlock(node, replaceMacros(node, macros, context)),
          resolveMacros,
          context
        );
      }
      if (ts.isSourceFile(node)) {
        return ts.visitEachChild(
          ts.updateSourceFileNode(node, replaceMacros(node, macros, context)),
          resolveMacros,
          context
        );
      }
      return ts.visitEachChild(node, resolveMacros, context);
    }
  };
  function fixMacros(
    node: ts.Block,
    context: ts.TransformationContext
  ): [ts.Expression | undefined, ts.Block] {
    const variableMap: Record<string, boolean> = {};
    let result: ts.Expression | undefined = undefined;
    const resultNode = ts.visitNode(node, visit);
    return [result, resultNode];
    function visit(node: ts.Node): ts.Node | undefined {
      if (ts.isReturnStatement(node)) {
        if (!node.expression) throw new Error("Expected macro to return value");
        result = ts.visitNode(node.expression, visit);
        return undefined;
      }
      if (ts.isPropertyAccessExpression(node)) {
        return ts.createPropertyAccess(
          ts.visitNode(node.expression, visit),
          node.name
        );
      }
      if (ts.isVariableDeclaration(node) && ts.isIdentifier(node.name)) {
        variableMap[node.name.text] = true;
      }
      if (ts.isIdentifier(node) && variableMap[node.text]) {
        return ts.createIdentifier(newName(node.text));
      }
      return ts.visitEachChild(node, visit, context);
      function newName(name: string) {
        return "__" + name.toString().toLocaleLowerCase() + counter.toString();
      }
    }
  }
  function replaceMacros(
    block: ts.BlockLike,
    macros: Macro[],
    context: ts.TransformationContext
  ): ts.Statement[] {
    let result: ts.Statement[] = [];
    for (const statement of block.statements) {
      const newStatement = ts.visitNode(statement, visit);
      result.push(newStatement);
    }
    return result;
    function visit(child: ts.Node): ts.Node {
      // TODO check if visit enought children
      if (ts.isBlock(child)) {
        return ts.createBlock(replaceMacros(child, macros, context));
      }
      const identifier = getIdentifier(child);
      if (identifier) {
        for (const { name, value } of macros) {
          if (identifier.text === name.text) {
            if (ts.isIdentifier(child)) {
              return value;
            }
            if (ts.isCallExpression(child)) {
              if (!ts.isArrowFunction(value)) {
                throw new Error("Expected function expression for macro value");
              }
              const newMacros = macros.slice();
              for (
                let i = 0;
                i < child.arguments.length && i < value.parameters.length;
                i++
              ) {
                const name = value.parameters[i].name;
                if (!ts.isIdentifier(name)) {
                  throw new Error(
                    "Expected identifier in macro function definition"
                  );
                }
                const arg = child.arguments[i];
                newMacros.push({ name, value: arg });
              }

              const block = ts.isBlock(value.body)
                ? value.body
                : ts.createBlock([ts.createReturn(value.body)]);
              counter++;
              const [resultName, resultBlock] = fixMacros(
                ts.createBlock(replaceMacros(block, newMacros, context)),
                context
              );
              result = result.concat(resultBlock.statements);
              if (!resultName) {
                throw new Error("Macro should return value");
              }
              return resultName;
            }
            throw new Error("Expected macro as call expression or identifier");
          }
        }
      }
      return ts.visitEachChild(child, visit, context);
    }
  }
}

function getIdentifier(node: ts.Node): ts.Identifier | undefined {
  if (ts.isCallExpression(node) && ts.isIdentifier(node.expression))
    return node.expression;
  if (ts.isIdentifier(node)) return node;
  return undefined;
}

export default transformer;
