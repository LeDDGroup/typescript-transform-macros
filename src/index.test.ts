import * as ts from "typescript";
import { resolve } from "path";
import transformer from "./";
import { readFileSync } from "fs";

describe("transformer", () => {
  it("should compile", () => {
    const inputFile = resolve(__dirname, "__fixtures/input.ts");
    const result = compileCode(readFileSync(inputFile).toString());
    expect(result).toMatchSnapshot();
  });
});

// function compile(file: string): string {
//   let content = "";
//   const program = ts.createProgram([file], {
//     target: ts.ScriptTarget.ESNext,
//     module: ts.ModuleKind.CommonJS
//   });
//   program.emit(
//     undefined,
//     (_, result) => (content = result),
//     undefined,
//     undefined,
//     {
//       after: [transformer(program)]
//     }
//   );
//   return content;
// }

function compileCode(source: string): string {
  let result = ts.transpileModule(source, {
    transformers: {
      before: [transformer()]
    },
    compilerOptions: {
      target: ts.ScriptTarget.ESNext,
      module: ts.ModuleKind.CommonJS
    }
  });

  return result.outputText;
}
