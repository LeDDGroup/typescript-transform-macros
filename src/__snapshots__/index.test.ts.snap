// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformer should compile 1`] = `
"const input_1 = [1, 2, 3, 4];
const length_1 = input_1.length;
const result_1 = [];
for (let i_1 = 0; i_1 < length_1; i_1++) {
    if (input_1[i_1]
        % 2 === 0)
        result_1.push(input_1[i_1]);
}
const input_2 = result_1;
const length_2 = input_2.length;
const result_2 = new Array(length_2);
for (let i_2 = 0; i_2 < length_2; i_2++) {
    result_2[i_2] =
        input_2[i_2]
            + 1;
}
const input_3 = result_2;
const length_3 = input_3.length;
const result_3 = new Array(length_3);
for (let i_3 = 0; i_3 < length_3; i_3++) {
    result_3[i_3] =
        input_3[i_3].toString();
}
console.log(result_3);
"
`;
